mobile_payment/gql_mutations.py

```diff
@@ -0,0 +1,421 @@
+        response = super().mutate_and_get_payload(root, info, **data)
+        try:
+            if info.context.user is AnonymousUser:
+                raise ValidationError(_("mutation.authentication_required"))
+            if not info.context.user.has_perms(MobilePaymentConfig.gql_mutation_create_transaction_perms):
+                raise PermissionDenied(_("unathorized"))
+            client_mutation_id = data.get("client_mutation_id")
+            transaction_uuid = data['uuid']
+            transacton = Transactions.objects.get(uuid=transaction_uuid )
+            if not transacton:
+                raise PermissionDenied (_("invalid transaction_uuid"))
+            transaction_id = transacton.transaction_id
+            otp = data['otp']
+            print(f' Transaction_id = {transaction_id}')
+            print (f'otp =  {otp}')
+            Qcell_process = process_request(otp, transaction_id)
+            if Qcell_process['responseCode'] == '1':
+                data['status'] = True
+                update_or_create_transaction(data, info)
+            else:
+                raise Exception(_("Failed to initiate payment"))
+            return ProcessTransactionMutation(internal_id=response.internal_id)
+        except Exception as exc:
+            logger.exception("transaction.mutation.failed_to_process_transaction")
+            return [{
+                    'message': _("transaction.mutation.failed_to_process_transaction"),
+                    'detail': str(exc)}]
+
+# *********************************** Afri Money ***********************************           
+class ProductInput(graphene.InputObjectType):
+    product_name = graphene.String(required=True)
+    product_code = graphene.String(required=True)
+    token = graphene.String(required = True)
+class ProcessPaymentInput(graphene.InputObjectType):
+    chf_id = graphene.String(required=True)
+    amount = graphene.Decimal(required=True)
+    transaction_id = graphene.String(required = False)
+    psp_service_provider_uuid = graphene.String(required = False)
+    policy = graphene.Field( ProductInput, required = False)
+    json_content = graphene.JSONString()
+
+def handle_payment(user, data):
+    chf_id = data['chf_id']
+    amount = data['amount']
+    psp_service_provider_uuid = data['psp_service_provider_uuid']
+    product_name = data['policy']['product_name']
+    product_code = data['policy']['product_code']
+    token = data['policy']['token']
+    json_content = json.dumps(data.get('json_content', []))
+    
+    
+
+    try:
+        insuree = Insuree.objects.filter(chf_id=chf_id, validity_to__isnull=True).first()
+        if not insuree:
+            return {"success": False, "message": "Please pass in the correct CH ID."}
+    except Insuree.DoesNotExist:
+        return {"success": False, "message": "Please pass in the correct CHF ID."}
+
+    
+    policy = Policy.objects.filter(family=insuree.family, uuid=token).first()
+    if policy:
```

Comment:
This logic is the same whatever the payment service provider. I expect it to be shared and common. I don't think this should be specific to one psp.

Reply:
No its not specific to one psp i will take a look at  it again

Reply:
are you referring to the token or what i don't understand

Reply:
okay but i don,t think am repeating my self in this code pls take a look again 
```
def handle_payment(user, data):
    chf_id = data['chf_id']
    amount = data['amount']
    psp_transaction_id = data['psp_transaction_id']
    psp_service_provider_uuid = data['psp_service_provider_uuid']
    product_name = data['policy']['product_name']
    product_code = data['policy']['product_code']
    token = data['policy']['token']
    audit_user_id = data['audit_user_id']
    json_content = json.dumps(data.get('json_ext', []))


    try:
        insuree = Insuree.objects.filter(chf_id=chf_id, validity_to__isnull=True).first()
        if not insuree:
            return {"success": False, "message": "Please pass in the correct CH ID."}
    except Insuree.DoesNotExist:
        return {"success": False, "message": "Please pass in the correct CHF ID."}

    
    policy = Policy.objects.filter(family=insuree.family, uuid=token, validity_to__isnull=True).first()
    if policy:
        product = policy.product
        if (
            amount == product.lump_sum
            and policy.status == Policy.STATUS_IDLE
            and product_name == product.name
            and product_code == product.code
        ):
            psp_service_provider = PaymentServiceProvider.filter_queryset().filter(uuid=psp_service_provider_uuid).first()
            transaction = PaymentTransaction.objects.create(amount=amount, payment_service_provider=psp_service_provider, status=1, psp_transaction_id = psp_transaction_id, json_content=json_content, datetime=date.today())
            transaction.save()
            premium = Premium.objects.create(amount=transaction.amount, policy=policy, pay_date=date.today(), pay_type="M", payment_transaction=transaction, audit_user_id =audit_user_id)
            premium.save()
            update_policy(premium, policy)
            return {'success': True}
        elif amount < product.lump_sum and policy.status == Policy.STATUS_IDLE and product_name == product.name and product_code == product.code:
            return {'success': False, 'message': "Amount is too small. It needs to be the exact amount of the product."}
        elif amount > product.lump_sum and policy.status == Policy.STATUS_IDLE and product_name == product.name and product_code == product.code:
            return {'success': False, 'message': "Amount is too large. pls enter the exact amount the product cost."}
        elif amount == product.lump_sum and policy.status != Policy.STATUS_IDLE and product_name == product.name and product_code == product.code:
            return {'success': False, 'message': "Please check again if your policy you are trying to pay for is idle  or active."}
    else:
        return {"success": False, "message": "Unable to process payment. insuree does not have any idle policy that requires payment."}
        
def update_policy(premium, policy):
    policy.status = Policy.STATUS_ACTIVE
    policy.effective_date = premium.pay_date if premium.pay_date > policy.start_date else policy.start_date
    policy.save()
```